### PR TITLE
catch the full of send/recv queue and resubmit

### PR DESCRIPTION
**Pull request by gipsyh**
Signed-off-by: gipsyh <gispyh.icu@gmail.com>

### CODE REVIEW

File: src/completion_queue.rs [hunk 1]
```diff
@@ -25,13 +26,13 @@ pub struct CompletionQueue {
 
 impl CompletionQueue {
     /// Get the internal cq ptr
-    pub(crate) const fn as_ptr(&self) -> *mut ibv_cq {
```
Comment (gipsyh) [inline src/completion_queue.rs:28]: CQ is not transparent for user, so pub(crate) is not necessary.

Comment (rogercloud) [reply src/completion_queue.rs:28]: Specific is better than implicit. So I recommend to keep `pub(crate)`.

File: src/completion_queue.rs [hunk 2]
```diff
@@ -46,18 +47,18 @@ impl CompletionQueue {
...
 
     /// Request notification on next complete event arrive
-    pub(crate) fn req_notify(&self, solicited_only: bool) -> io::Result<()> {
+    pub fn req_notify(&self, solicited_only: bool) -> io::Result<()> {
         let errno = unsafe {
             ibv_req_notify_cq(self.inner_cq.as_ptr(), if solicited_only { 1 } else { 0 })
         };
         if errno != 0_i32 {
-            return Err(io::Error::from_raw_os_error(errno));
+            return Err(io::Error::from_raw_os_error(errno.mul(-1_i32)));
```
Comment (rogercloud) [inline src/completion_queue.rs:55]: is `0 - errno`  better?

File: src/completion_queue.rs [hunk 3]
```diff
@@ -115,16 +116,17 @@ pub struct WorkCompletion {
...
 impl WorkCompletion {
     /// Get work request Id
-    pub(crate) const fn wr_id(&self) -> WorkRequestId {
+    pub const fn wr_id(&self) -> WorkRequestId {
         WorkRequestId(self.inner_wc.wr_id)
     }
 
     /// Get work completion result, if success returns length, otherwise returns error
-    pub(crate) fn err(&self) -> Result<usize, WCError> {
+    #[allow(clippy::unwrap_in_result, clippy::expect_used)]
```
Comment (rogercloud) [inline src/completion_queue.rs:124]: We don't use unwrap and expect in lib crate, as people are not expecting panic while using these crates.

File: src/queue_pair.rs [hunk 4]
```diff
@@ -1,551 +1,581 @@
...
+        match s.state {
+            QueuePairOpsState::Init => {
+                let (wr_id, recv) = s.qp.event_listener.register();
+                s.state = QueuePairOpsState::Submit(wr_id, Some(recv));
+                Pin::new(s).poll(cx)
+            }
+            QueuePairOpsState::Submit(wr_id, ref mut recv) => {
+                if let Err(e) = s.op.submit(&s.qp, wr_id) {
+                    if s.op.should_resubmit(&e) {
+                        let sleep = Box::pin(sleep(Duration::from_micros(10)));
```
Comment (rogercloud) [inline src/queue_pair.rs:558]: Put the delay number to a const.

File: src/completion_queue.rs (hunk 1 above)
Comment (gipsyh) [reply src/completion_queue.rs:28]: https://github.com/datenlord/async-rdma/pull/28/commits/61644b432d838cfa16feb63b2e17b222d4125b17 fixed

File: src/completion_queue.rs (hunk 2 above)
Comment (gipsyh) [reply src/completion_queue.rs]: https://github.com/datenlord/async-rdma/pull/28/commits/61644b432d838cfa16feb63b2e17b222d4125b17 fixed

File: src/queue_pair.rs (hunk 4 above)
Comment (gipsyh) [reply src/queue_pair.rs]: https://github.com/datenlord/async-rdma/pull/28/commits/d59d1679f0aec96edfdfe5735abb0108a2ac7dd3 fixed

File: src/completion_queue.rs (hunk 3 above)
Comment (gipsyh) [reply src/completion_queue.rs]: https://github.com/datenlord/async-rdma/pull/28/commits/97b18215ebd842124b08e823f0dd9b08f4da5da2 fixed

File: src/completion_queue.rs (hunk 1 above)
Comment (gipsyh) [reply src/completion_queue.rs]: https://github.com/datenlord/async-rdma/pull/28/commits/e9f2f1b34afc6a9bfc4d0e0a8441a42890c80b36 fixed

**Review by rogercloud — DISMISSED** [review]

**Review by rogercloud — APPROVED** [review]

**Pull request merged**